### PR TITLE
Schema gen

### PR DESCRIPTION
**Pull request by nickav**

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -63,13 +65,36 @@ func (c *Client) Wait() error {
...
 	client.System = &SystemModule{client: client}
 	client.App = &AppModule{client: client}
 	client.Menu = &MenuModule{client: client}
 	client.Shell = &ShellModule{client: client}
+
+	values := reflect.ValueOf(*client.Shell)
+	types := values.Type()
+	for i := 0; i < values.NumField(); i++ {
+		field := types.Field(i)
+		log.Println(field.Index[0], field.Name, reflect.TypeOf(field.Type).Kind() == reflect.Func)
```
Comment (nickav) [inline client/client.go:89]: wasn't sure how to check if this field is a function, this expression always returns false

File: schema/main/main.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		// - resource.Handle gets replaced with Handle
+		//{"bridge/api/menu/menu.go", "client/menu.go", "Menu"},
+
+		// @Incomplete:
+		// ShellModule is missing: OnShortcut func(event Event)
+		{"bridge/api/shell/shell.go", "client/shell.go", "Shell"},
+		{"bridge/api/system/system.go", "client/system.go", "System"},
+
+		//{"bridge/api/window/window.go", "client/window.go", "Window"},
+	}
```
Comment (nickav) [inline schema/main/main.go:39]: comments describe differences between client code and bridge code

File: client/client.go (hunk 1 above)
Comment (progrium) [reply client/client.go:89]: TypeOf returns the type, so `reflect.TypeOf(field.Type)` is redundant. You can just do `field.Type.Kind()`

File: client/client.go [hunk 3]
```diff
@@ -63,13 +65,36 @@ func (c *Client) Wait() error {
 	return c.cmd.Wait()
 }
 
+func Bind(client *Client, name string, p interface{}) {
+	f := reflect.ValueOf(p).Elem()
+	f.Set(reflect.MakeFunc(f.Type(), func(args []reflect.Value) (result []reflect.Value) {
+		ctx, _ := reflect.ValueOf(args[0]).Interface().(context.Context)
+		_, err := client.Call(ctx, name, fn.Args{args[1]}, nil)
+		return []reflect.Value{reflect.ValueOf(err).Elem()}
```
[the remark this answers is not in the document]
Comment (progrium) [reply client/client.go]: There are some helper functions in https://github.com/progrium/qtalk-go/blob/main/fn/call.go that should be helpful here.

[the remark this answers is not in the document]
Comment (progrium) [reply client/client.go]: We can also assume that the function will always return an error or a value and an error. If the remote function doesn't return an error, we always want to anyway for rpc/net errors. The case for even more return values we can ignore for now.

File: bridge/api/system/system.go [hunk 4]
```diff
@@ -27,10 +27,10 @@ type Position = misc.Position
 
 type Size = misc.Size
 
-func (m module) Displays() []Display {
+func (m *module) Displays() []Display {
```
Comment (progrium) [inline bridge/api/system/system.go:30]: not sure this is a good idea. why was this added?

File: bridge/api/shell/shell.go [hunk 5]
```diff
@@ -6,7 +6,9 @@ func init() {
 	Module = &module{}
 }
 
-type module struct{}
+type module struct {
+	OnShortcut func(event Event)
```
Comment (progrium) [inline bridge/api/shell/shell.go:10]: I'm assuming these were added so they'd be generated? But these are manual additions and have no use here in the bridge, so they should not be added.

File: schema/out/shell.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package client
+
+import (
+	"context"
+)
+
+type ShellModule struct {
+	client *Client
```
Comment (progrium) [inline schema/out/shell.go:8]: also something that will be added manually. this struct should just have the API functions.

File: bridge/api/system/system.go (hunk 4 above)
Comment (nickav) [reply bridge/api/system/system.go:30]: just to make it more consistent with other APIs / parsing slightly easier, but why would this possibly be bad? don't feel strongly either way

File: bridge/api/shell/shell.go (hunk 5 above)
Comment (nickav) [reply bridge/api/shell/shell.go:10]: yeah, how would we handle that in general for parsing?

Comment (progrium) [reply bridge/api/shell/shell.go:10]: I don't follow. Handle what?

File: bridge/api/system/system.go (hunk 4 above)
Comment (progrium) [reply bridge/api/system/system.go:30]: Looks like you're right, I assumed it already was consistent. Parsing should handle either case though. I believe the receiver was not a pointer because there was no state written back to it, so it didn't need to be. It's bad going the other way, but in this case it doesn't matter.

Comment (nickav) [reply bridge/api/system/system.go:30]: the parser handles both cases so whichever you prefer is good with me!

File: bridge/api/shell/shell.go (hunk 5 above)
Comment (nickav) [reply bridge/api/shell/shell.go:10]: how would you dynamically add the callback functions to each struct? e.g. `ShellModule` needs `OnShortcut`

Comment (progrium) [reply bridge/api/shell/shell.go:10]: Like I said, that would not be automated.

File: bridge/api/menu/menu.go [hunk 7]
```diff
@@ -7,7 +7,9 @@ import (
 
 var Module *module
 
-type module struct{}
+type module struct {
+	OnClick func(event Event)
```
Comment (progrium) [inline bridge/api/menu/menu.go:11]: missed this one.

**Comment by progrium** [comment]
great, nice work
